src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -57,20 +57,27 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
         private readonly Lazy<IVsPackageUninstaller> _packageUninstaller;
         private readonly Lazy<IVsPackageSourceProvider> _packageSourceProvider;
 
-        private JoinableTask<ImmutableArray<PackageSource>?> _packageSourcesAsync;
         private IVsPackage _nugetPackageManager;
 
-        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
-
         // We keep track of what types of changes we've seen so we can then determine what to
         // refresh on the UI thread.  If we hear about project changes, we only refresh that
         // project.  If we hear about a solution level change, we'll refresh all projects.
-        private bool _solutionChanged;
-        private readonly HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+        private AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId changedProject)> _workQueue;
```

Comment:
switched to the work-queue model which already handles throttling and processing of large streams of requests.